src/util/check.h

```diff
@@ -46,15 +46,26 @@ class NonFatalCheckError : public std::runtime_error
 #error "Cannot compile without assertions!"
 #endif
 
+/** Helper for Assert() */
+void assertion_fail(const char* file, int line, const char* func, const char* assertion);
+
 /** Helper for Assert() */
 template <typename T>
-T get_pure_r_value(T&& val)
+inline T&& inline_assertion_check(T&& val, const char* file, int line, const char* func, const char* assertion)
 {
+    if (!val) assertion_fail(file, line, func, assertion);
     return std::forward<T>(val);
 }
 
 /** Identity function. Abort if the value compares equal to zero */
-#define Assert(val) ([&]() -> decltype(get_pure_r_value(val)) { auto&& check = (val); assert(#val && check); return std::forward<decltype(get_pure_r_value(val))>(check); }())
+#define Assert(val) inline_assertion_check(val, __FILE__, __LINE__, __func__, #val)
+
+/** Helper for Assume() */
+template <typename T>
+inline T&& skip_assumption_check(T&& val)
+{
+    return std::forward<T>(val);
+}
```

Comment:
It might be possible to remove the separate helpers and just use the same function for both:

```diff
diff --git a/src/util/check.h b/src/util/check.h
index 3fa49c0a3e..4c2ee4501b 100644
--- a/src/util/check.h
+++ b/src/util/check.h
@@ -51,21 +51,22 @@ void assertion_fail(const char* file, int line, const char* func, const char* as
 
 /** Helper for Assert() */
 template <typename T>
-inline T&& inline_assertion_check(T&& val, const char* file, int line, const char* func, const char* assertion)
+T&& maybe_abort_on_check(bool IS_ASSERT, T&& val, const char* file, int line, const char* func, const char* assertion)
 {
-    if (!val) assertion_fail(file, line, func, assertion);
+    if (!val) {
+        if (IS_ASSERT // Abort inside Assert()
+#ifdef ABORT_ON_FAILED_ASSUME
+            || true // Abort inside Assume()
+#endif
+        ) {
+            assertion_fail(file, line, func, assertion);
+        }
+    }
     return std::forward<T>(val);
 }
 
 /** Identity function. Abort if the value compares equal to zero */
-#define Assert(val) inline_assertion_check(val, __FILE__, __LINE__, __func__, #val)
-
-/** Helper for Assume() */
-template <typename T>
-inline T&& skip_assumption_check(T&& val)
-{
-    return std::forward<T>(val);
-}
+#define Assert(val) maybe_abort_on_check(true, val, __FILE__, __LINE__, __func__, #val)
 
 /**
  * Assume is the identity function.
@@ -77,10 +78,6 @@ inline T&& skip_assumption_check(T&& val)
  * - For non-fatal errors in interactive sessions (e.g. RPC or command line
  *   interfaces), CHECK_NONFATAL() might be more appropriate.
  */
-#ifdef ABORT_ON_FAILED_ASSUME
-#define Assume(val) Assert(val)
-#else
-#define Assume(val) skip_assumption_check(val)
-#endif
+#define Assume(val) maybe_abort_on_check(false, val, __FILE__, __LINE__, __func__, #val)
 
 #endif // BITCOIN_UTIL_CHECK_H

Reply:
Made it a template argument instead of a function param, used `if constexpr` and left the `!val` as the inner `if`, but otherwise taken up.